sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -514,3 +514,178 @@ case class StringToMap(text: Expression, pairDelim: Expression, keyValueDelim: E
+          s"at even position.")
+    } else if (valExprs.contains(null)) {
+      TypeCheckResult.TypeCheckFailure(
+        s"Only non-null expressions are allowed to appear at odd positions after first position.")
+    } else {
+      TypeCheckResult.TypeCheckSuccess
+    }
+  }
+
+  override def eval(input: InternalRow): Any = {
+    val structValue = struct.eval(input)
+    if (structValue == null) {
+      null
+    } else {
+      val existingValues: Seq[(FieldName, Any)] =
+        ogStructType.fieldNames.zip(structValue.asInstanceOf[InternalRow].toSeq(ogStructType))
+      val addOrReplaceValues: Seq[(FieldName, Any)] =
+        pairs.map { case (fieldName, expression) => (fieldName, expression.eval(input)) }
+      val newValues = loop(existingValues, addOrReplaceValues).map(_._2)
+      InternalRow.fromSeq(newValues)
+    }
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val structGen = struct.genCode(ctx)
+    val addOrReplaceFieldsGens = valExprs.map(_.genCode(ctx))
+    val resultCode: String = {
+      val structVar = structGen.value
+      type NullCheck = String
+      type NonNullValue = String
+      val existingFieldsCode: Seq[(FieldName, (NullCheck, NonNullValue))] =
+        ogStructType.fields.zipWithIndex.map {
+          case (structField, i) =>
+            val nullCheck = s"$structVar.isNullAt($i)"
+            val nonNullValue = CodeGenerator.getValue(structVar, structField.dataType, i.toString)
+            (structField.name, (nullCheck, nonNullValue))
+        }
+      val addOrReplaceFieldsCode: Seq[(FieldName, (NullCheck, NonNullValue))] =
+        fieldNames.zip(addOrReplaceFieldsGens).map {
+          case (fieldName, fieldExprCode) =>
+            val nullCheck = fieldExprCode.isNull.code
+            val nonNullValue = fieldExprCode.value.code
+            (fieldName, (nullCheck, nonNullValue))
+        }
+      val newFieldsCode = loop(existingFieldsCode, addOrReplaceFieldsCode)
+      val rowClass = classOf[GenericInternalRow].getName
+      val rowValuesVar = ctx.freshName("rowValues")
+      val populateRowValuesVar = newFieldsCode.zipWithIndex.map {
+        case ((_, (nullCheck, nonNullValue)), i) =>
+          s"""
+             |if ($nullCheck) {
+             | $rowValuesVar[$i] = null;
+             |} else {
+             | $rowValuesVar[$i] = $nonNullValue;
+             |}""".stripMargin
+      }.mkString("\n|")
+
+      s"""
+         |Object[] $rowValuesVar = new Object[${dataType.length}];
+         |
+         |${addOrReplaceFieldsGens.map(_.code).mkString("\n")}
+         |$populateRowValuesVar
+         |
+         |${ev.value} = new $rowClass($rowValuesVar);
+          """.stripMargin
+    }
+
+    if (nullable) {
+      val nullSafeEval =
+        structGen.code + ctx.nullSafeExec(struct.nullable, structGen.isNull) {
+          s"""
+             |${ev.isNull} = false; // resultCode could change nullability.
```

Comment:
plz wrap the comment with `CodeGenerator.registerComment`

Reply:
For my understanding, could you please explain how doing this would help? 
For reference, I was just copying the style of https://github.com/apache/spark/blob/b28c76bd7ad7727c184bc1a8e9d4c28d85389ac3/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala#L602